rnnalyse/extractors/base_extractor.py

```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
               f'Time: {minutes:>3.0f}m {seconds:>2.2f}s\t'
               f'Speed: {speed:.2f}s/sen')
 
-    def _extract_sentence(self, sentence: Sentence) -> None:
+    def _extract_sentence(self, sentence: LabeledSentence) -> int:
         """ Generates the embeddings of a sentence and writes to file.
 
         Parameters
         ----------
         sentence : Sentence
             The to-be-extracted sentence, represented as a list of strings.
+
+        Returns
+        -------
+        num_extracted: int
+            Number of extracted activations for this sentence.
         """
+
+        sen_activations: PartialActivationDict = self._init_sen_activations(0)
+
+        activations: FullActivationDict = self.init_lstm_states.states
+
+        for i, token in enumerate(sentence.sen):
+            out, activations = self.model(token, activations)
+
+            # Check whether current activations match criterion defined in selection_func
+            if self.selection_func(i, token, sentence):
+                for l, name in self.activation_names:
+                    sen_activations[(l, name)] = np.append(
+                        sen_activations[(l, name)], activations[l][name].detach().numpy()[np.newaxis, ...], axis=0
+                    )
+
+        num_extracted = list(sen_activations.values())[0].shape[0]
+        self._dump_activations(sen_activations)
+
+        return num_extracted
+
+    def extract_average_eos_activations(self):
+        """ Extract average end of sentence activations and dump them to a file. """
+
         def _incremental_avg(old_avg, new_value):
             # n_sens only get incremented after _extract_sentence is called, therefore + 1 here
             return old_avg + 1 / (self.n_sens + 1) * (new_value - old_avg)
 
-        sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
+        # Init states
+        avg_eos_states = {
+            l: {'hx': torch.zeros(self.model.hidden_size), 'cx': torch.zeros(self.model.hidden_size)}
```

Comment:
It took me a while to figure out why we need the `l`. Maybe it would be more obviously clear if we renamed it `layer`?

Reply:
I was sticking to @jumelet's notation. As you know I like to write things out :-P

Reply:
Hahaha I was already surprised indeed. I like to optimise making things understandable and keeping lines short. E.g., I don't mind using abbreviations in certain contexts, such as act_len instead of activation_length, or use single characters for small loops for which the meaning is not that important, but I'm also for helping the code reader as much as possible.

I'm sure we can figure out some style that makes us all happy :)